File: C++/3.无重复字符的最长子串.cpp
/*
 * @lc app=leetcode.cn id=3 lang=cpp
 *
 * [3] 无重复字符的最长子串
 */

// @lc code=start
class Solution
{
public:
    int lengthOfLongestSubstring(string s)
    {
        int ans = 0, left = 0, len = s.length();
        int last[255];
        memset(last, -1, sizeof last);

        for (int i = 0; i < len; i++)
        {
            if (last[s[i]] >= left)
                left = last[s[i]] + 1;
            last[s[i]] = i;
            ans = max(ans, i - left + 1);
        }
        return ans;
    }
};
// @lc code=end

File: JavaScript/3.无重复字符的最长子串.js
/*
 * @lc app=leetcode.cn id=3 lang=javascript
 *
 * [3] 无重复字符的最长子串
 */

// @lc code=start
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function (s) {
  const last = Array(255).fill(-1);
  let ans = 0;
  let left = 0;

  for (let i = 0; i < s.length; i++) {
    const charCode = s.charCodeAt(i);

    if (last[charCode] >= left) {
      left = last[charCode] + 1;
    }

    last[charCode] = i;
    ans = Math.max(ans, i - left + 1);
  }

  return ans;
};
// @lc code=end
